src/trace/ord/merge_batcher.rs

```diff
@@ -0,0 +1,405 @@
+        let mut result = self.empty();
+
+        let mut list1 = VecQueue::from(list1);
+        let mut list2 = VecQueue::from(list2);
+
+        let mut head1 = if !list1.is_empty() {
+            VecQueue::from(list1.pop())
+        } else {
+            VecQueue::new()
+        };
+        let mut head2 = if !list2.is_empty() {
+            VecQueue::from(list2.pop())
+        } else {
+            VecQueue::new()
+        };
+
+        // while we have valid data in each input, merge.
+        while !head1.is_empty() && !head2.is_empty() {
+            while (result.capacity() - result.len()) > 0 && !head1.is_empty() && !head2.is_empty() {
+                let cmp = {
+                    let x = head1.peek();
+                    let y = head2.peek();
+                    (&x.0).cmp(&y.0)
+                };
+                match cmp {
+                    Ordering::Less => unsafe {
+                        push_unchecked(&mut result, head1.pop());
+                    },
+                    Ordering::Greater => unsafe {
+                        push_unchecked(&mut result, head2.pop());
+                    },
+                    Ordering::Equal => {
+                        let (data1, mut diff1) = head1.pop();
+                        let (_data2, diff2) = head2.pop();
+                        diff1.add_assign_by_ref(&diff2);
+                        if !diff1.is_zero() {
+                            unsafe {
+                                push_unchecked(&mut result, (data1, diff1));
+                            }
+                        }
+                    }
+                }
+            }
+
+            if result.capacity() == result.len() {
+                output.push(result);
+                result = self.empty();
+            }
+
+            if head1.is_empty() {
+                let done1 = head1.done();
+                if done1.capacity() == Self::buffer_size() {
+                    self.stash.push(done1);
+                }
+                head1 = if !list1.is_empty() {
+                    VecQueue::from(list1.pop())
+                } else {
+                    VecQueue::new()
+                };
+            }
+            if head2.is_empty() {
+                let done2 = head2.done();
+                if done2.capacity() == Self::buffer_size() {
+                    self.stash.push(done2);
+                }
+                head2 = if !list2.is_empty() {
+                    VecQueue::from(list2.pop())
+                } else {
+                    VecQueue::new()
+                };
+            }
+        }
+
+        if !result.is_empty() {
+            output.push(result);
+        } else if result.capacity() > 0 {
+            self.stash.push(result);
+        }
+
+        if !head1.is_empty() {
+            let mut result = self.empty();
+            for _ in 0..head1.len() {
+                result.push(head1.pop());
+            }
```

Comment:
Shouldn't this be an `.extend()` call or something? Plus we can pre-allocate `result`, we know its exact size